rewrite matplotlib section lesson 08

stijnvanhoey opened the pull request:
This PR aims to change/improve the matplotlib section of the `Data Ingest and Visualization - Matplotlib and Pandas` lesson, providing the following changes:
- focus on the link between matplotlib, plotnine(ggplot) and Pandas instrad of *learning this new library*
- add a section and challenge about saving plots, as requested in #293
- it overcomes the issue of opening and closing plots, see #277, by focusing on the object oriented approach of matplotlib

It indeed basically excludes the whole introduction of the stateful approach (cfr. Matlab style), but I do think this is an improvement, as:
- the link with the other episodes is made in the new version. The main concept is: start working in Pandas/plotnine, but be able to further customize with matplotlib when required.
- the object oriented way is more powerful for adjustments and using the TAB, people can really explore the structure of the object
- it is a bit strange to start new concepts (creating plots from scratch in matplotlib) in a 'putting it together( episode.

Still, this PR is a suggestion and I'm certainly open to comments, other ideas,... It was just easier to provide the PR than trying to explain the concept. I also realize that does not complete reacts to the discussion in #254, but I do think the matplotlib-section can still be integrated in a real capstone project ad maybe this can (re)open the discussion on what to do with this episode.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
-    - "FIXME"
+	- "Matplotlib is the engine behind plotnine and Pandas plots."
+	- "The object-oriented way of matplotlib makes detailed customization possible."
+	- Use `savefig` method to save a figure as a file;
```
use double quotes. remote semicolon at the end

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
-toolbox `matplotlib.pyplot` is a collection of functions that make matplotlib
-work like MATLAB. In most cases, this is all that you will need to use, but
-there are many other useful tools in matplotlib that you should explore.
+[Matplotlib](http://matplotlib.org/) is a Python package used widely throughout the scientific Python community to produce high quality and publication-ready graphics. It supports a wide range of output formats including PNG (and other raster formats), PostScript/EPS, PDF and SVG. 
```
Title says `Matplotlib plot *library*`, the first sentence says ...`Python *package*`. It *might* be useful to use one or the other but not both.

maxim-belkin replied:
`high-quality` because it refers to `graphics`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
 
-We will cover a few basic commands for formatting plots in this lesson. A great
-resource for help styling your figures is the matplotlib gallery
+At the same time, matplotlib is the actual engine behind the plotting capabilities of both the Pandas as well as the plotnine package. So, if you you call the `.plot` functionality of Pandas, as we did in the previous episodes, you actually used the matplotlib package:
```
`as well as` ~> `and` (this is my personal preference, so you can ignore it if you disagree)

maxim-belkin replied:
Fix: `So, if you call....` and `you actually used` (either: `When you called` or `you actually use`)

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
+
+~~~
+    import pandas as pd
+    surveys = pd.read_csv("./data/surveys.csv")
```
remove whitespaces in the beginning

maxim-belkin replied:
of the first two lines

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
+~~~
+{: .language-python}
+
+![png](../fig/08_scatter_surveys.png)
```
provide a better description than `png`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
+
+The returned object is a `matplotlib.axes._subplots.AxesSubplot` matplotlib object (check it yourself with `type(my_plot)`) and the power of matplotlib is available to further adjust these plots as it is created with matplotlib itself.
+
+> ~~~
```
remove `~~~`, add an H2 title

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
+>     do as much as you easily can in a convenience layer, i.e. start
+>     creating the plot in Pandas or plotnine, and then use matplotlib
+>     for the rest.
+> ~~~
```
remove whitespaces in the beginnings of lines
remove `~~~`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
+>     creating the plot in Pandas or plotnine, and then use matplotlib
+>     for the rest.
+> ~~~
+> {: .callout}
```
`> {: .callout}` -> `{: .callout}`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
+> {: .callout}
+
+We will cover a few basic commands for creating and formatting plots with matplotlib in this lesson. A great resource for help creating and styling your figures is the matplotlib gallery
 (http://matplotlib.org/gallery.html), which includes plots in many different
```
add `< >` around the URL

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,32 +89,37 @@ also change the name of the columns in the DataFrame like this:
 {: .language-python}
 
-## Make a line plot of your data
+## Matplotlib plot library
```
Use either `matplotlib plotting library` or `matplotlib library`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -129,166 +137,171 @@ writing:
+    import numpy as np
+    x = np.linspace(0, 5, 10)
+    y = x ** 2
 ~~~
```
remove whitespaces in the beginnings of lines

maxim-belkin replied:
in `python-novice-inflammation` we agreed to not import `numpy` as `np`. I'd suggest this lesson does the same (unnecessary cognitive load)

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -129,166 +137,171 @@ writing:
 ~~~
-    plt.plot([6.8, 4.3, 3.2, 8.1], list_numbers)
-    plt.show()
+    plt.plot(x, y, '-')
```
same as above.. remove whitespaces in the beginning of the line

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -129,166 +137,171 @@ writing:
-line. For example, we can make the line red (`'r'`), with circles at every data
-point (`'o'`), and a dot-dash pattern (`'-.'`). Look through the matplotlib
-gallery for more examples.
+![png](../fig/08_line_plot.png)
```
provide better description than `png`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -129,166 +137,171 @@ writing:
-gallery for more examples.
+![png](../fig/08_line_plot.png)
+
+or create a figure and ax object first and add the plot to the created ax object:
```
ax?

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -129,166 +137,171 @@ writing:
-    plt.plot([6.8, 4.3, 3.2, 8.1], list_numbers, 'ro-.')
-    plt.axis([0,10,0,6])
-    plt.show()
+    fig, ax = plt.subplots()  # initiate an empty figure and ax matplotlib object
```
remove whitespace before `fig`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -129,166 +137,171 @@ writing:
-A single figure can include multiple lines, and they can be plotted using the
-same `plt.plot()` command by adding more pairs of x values and y values (and
-optionally line styles):
+Although the latter requires a little bit more code to create the same plot, the advantage is that we now have **full control** of where the plot axes are placed, and we can easily add new items or, for example more than one axis to the figure and adapting the labels::
```
`::` ~> `:`
`or, for example, more than`...
`adapting` ~> `adapt` ?

stijnvanhoey commented:
Thanks for the detailed review, the comments on the 'template styling' helped me a lot! I hope I addressed all the comments, feel free to provide more feedback.

wrightaprilm commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,218 @@ also change the name of the columns in the DataFrame like this:
 ~~~
-    import matplotlib.pyplot as plt
+import pandas as pd
+surveys = pd.read_csv("./data/surveys.csv")
```
previous lessons just do data/surveys.csv, would you mind switching to that?

stijnvanhoey replied:
consistency matter, I'll fix it

wrightaprilm commented:
Sorry I'm so late in getting to this. I think once Maxim's notes are addressed, we're good to go. Thanks for all your effort on this, it looks really nice.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,218 @@ also change the name of the columns in the DataFrame like this:
 
-    plt.legend(loc='upper left', shadow=True, fontsize='x-large')
+~~~
+fig, ax1 = plt.subplots() #prepare a matplotlib figure
```
add a space after `#`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,218 @@ also change the name of the columns in the DataFrame like this:
-    # This is a second figure
-    plt.figure(2)
-    plt.plot(t, t**2, 'bs-', label='square')
+fig, ax1 = plt.subplots() #prepare a matplotlib figure
```
add a space after `#`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,218 @@ also change the name of the columns in the DataFrame like this:
-    plt.legend(loc='upper left', shadow=True, fontsize='x-large')
-    plt.title('This is figure 2')
+surveys.plot("hindfoot_length", "weight", 
+kind="scatter", ax=ax1) # use Pandas for plotting
```
```
surveys.plot("hindfoot_length", "weight", kind="scatter", ax=ax1)
```

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,218 @@ also change the name of the columns in the DataFrame like this:
+p9_ax.set_xlabel("Hindfoot length")
+p9_ax.tick_params(labelsize=16, pad=8)
+p9_ax.set_title('Scatter plot of weight versus hindfoot length', fontsize=15)
+my_plt_version
```
`my_plt_version.show()`?

stijnvanhoey replied:
Is not required when using the matplotlib inline magic function as far as I know...

wrightaprilm replied:
This works fine on my install of jupyter notebook.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,218 @@ also change the name of the columns in the DataFrame like this:
-2 dimensional line plots. Look through the examples in
-http://matplotlib.org/users/screenshots.html and try a few of them (click on the
+Matplotlib can make many other types of plots in much the same way that it makes 2 dimensional line plots. Look through the examples in
+<http://matplotlib.org/users/screenshots.html and try a few of them (click on the
```
```
<http://matplotlib.org/users/screenshots.html>
```

stijnvanhoey commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,218 @@ also change the name of the columns in the DataFrame like this:
-    list_numbers = [1.5, 4, 2.2, 5.7]
-    plt.plot(list_numbers)
-    plt.show()
+%matplotlib inline
```
Maybe this has changed in a new version of it is different in Jupyter lab, but I would keep it for now.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,216 @@ also change the name of the columns in the DataFrame like this:
-function `plot()` receives two lists, it assumes the first one is the x-values
-and the second the y-values. The line connecting the points will follow the list
-in order:
+Matplotlib integrates nicely with the `numpy` package and can use numpy arrays as input of the available plot functions. Consider the following example data, created with `numpy`:
```
👍

maxim-belkin commented:
@wrightaprilm, Now that the PR reads well, I suggest we discuss if it is good "as is" in its entirety, or if we'd like to keep some parts of the old text in.

wrightaprilm commented:
I like this a lot, in particular with respect to some of the larger discussions we had about matplot vs. ggplot vs. plotnine in #246. I'd like to invite @katrintirok to take a peek at these changes, as well. I'm pretty happy with them, as-is, and I think they help the lesson strike a nice balance between the GoG framework and the matplot framework.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,216 @@ also change the name of the columns in the DataFrame like this:
-creates an object in memory but doesn't produce a visible plot. The ipython
-notebooks (if using `%matplotlib inline`) will automatically show you the figure
-even if you don't write `plt.show()`, but get in the habit of including this
-command!
```
`.show()` was previously mentioned and was, in fact, recommended. I'm not a big Jupyter user (`vi` habits take precedence), so I'd argue that it has to be at least mentioned.

wrightaprilm replied:
Coming back from maternity leave this week, so I'm getting back at looking at the repo. Is the change requested here to make clear that .show() should be included for portability? If this is the last thing standing between merging and not, I can make this change. I just want to make sure I understand the objection.

maxim-belkin replied:
those who don't use Jupyter notebooks will not see anything without the `.show()` command. While I appreciate all the benefit that Jupyter Notebook brings, I feel it is important to emphasize the fact that it does some things behind the scenes that learners will have to remember to do themselves in any other environment.

wrightaprilm replied:
OK, I see the issue. I can fix that (as an also Vim-defaulter). I can merge this (since it fixes several things in our issue queue) and add the language in a separate commit. Does that sound all right?

maxim-belkin replied:
You can add changes directly to this branch. I'll send you a message in Slack.

stijnvanhoey replied:
Good point @maxim-belkin I would indeed keep this text there to make sure non-notebook users are having the required info.

wrightaprilm commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -158,6 +159,15 @@ plt.plot(x, y, '-')
 or create a matplotlib `figure` and `axis` object first and add the plot to the created objects:
 
+> ## Tip: Cross-Platform Visualization of Figures
+> Jupyter Notebooks make many aspects of data analysis and visualization very easy. This includes
```
LOL, edited to reflect easy vs. simpl_er_

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-The function `legend()` adds a legend to the figure, and the optional keyword
-arguments change its style. By default [typing just `plt.legend()`], the legend
-is on the upper right corner and has no shadow.
+![png](../fig/08_line_plot.png)
```
I'd suggest providing a better description for the figure than `png` :)

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
     - "Why should I use Python to create plots?"
 objectives:
     - "Import the pyplot toolbox to create figures in Python."
+    - "Adapt maptlotlib objects created by Pandas or plotnine."
```
> Adapt ... objects

might be not super clear. Perhaps, `Learn how to change / adjust plots... ` or something like that

wrightaprilm replied:
Edited to better reflect role of matplotlib

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
+    - "Use matplotlib to make further adjustments to Pandas or plotnine objects."
 keypoints:
-    - "FIXME"
+    - "Matplotlib is the engine behind plotnine and Pandas plots."
```
How about: `Pandas and plotnine use matplotlib to create plots` ?

maxim-belkin replied:
I, actually, like the original version of this keypoint. So, my suggestion here is only a suggestion.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
 keypoints:
-    - "FIXME"
+    - "Matplotlib is the engine behind plotnine and Pandas plots."
+    - "The object-oriented way of matplotlib makes detailed customization possible."
```
I think keypoints have to be more specific `Plots created with Pandas and plotnine can be customized with matplotlib methods` ?

wrightaprilm replied:
As a replacement for the second key point? I like that.

stijnvanhoey replied:
I like the suggestion as well

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
     - "Why should I use Python to create plots?"
 objectives:
     - "Import the pyplot toolbox to create figures in Python."
+    - "Use matplotlib to make further adjustments to Pandas or plotnine objects."
```
I'd suggest removing `further`

stijnvanhoey replied:
fine for me

wrightaprilm commented:
I think we've zapped all the language issues. I'm happy for merge if everyone else is.

maxim-belkin commented:
The keypoint `The object-oriented way of matplotlib...` is still cryptic for those who're new to programming.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
-    - "FIXME"
+    - "Matplotlib is the engine behind plotnine and Pandas plots."
+    - "The object-oriented way of matplotlib makes detailed customization possible."
+    - "Use `savefig` method to save a figure as a file."
```
How about something like `Export figure to an image using `savefig` method` ?

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-    %matplotlib inline
-~~~
-{: .language-python}
+The returned object is a matplotlib object (check it yourself with `type(my_plot)`) and the power of matplotlib is available to further adjust these plots as it is created with matplotlib itself.
```
> The returned object, `my_plot`, is a matplotlib _object_

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-are just using a list as an example!):
+> ## Tip
+> Matplotlib itself can be overwhelming, so a useful strategy is to 
+> do as much as you easily can in a convenience layer, i.e. start
```
i.e. -> _i.e._

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-    plt.show()
-~~~
-{: .language-python}
+We will cover a few basic commands for creating and formatting plots with matplotlib in this lesson. A great resource for help creating and styling your figures is the matplotlib gallery
```
Need to break this like to not go over the 100-character limit

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-{: .language-python}
+We will cover a few basic commands for creating and formatting plots with matplotlib in this lesson. A great resource for help creating and styling your figures is the matplotlib gallery
+(<http://matplotlib.org/gallery.html>), which includes plots in many different
+styles and the source code that creates them. 
```
> source codes used to create them.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-line. For example, we can make the line red (`'r'`), with circles at every data
-point (`'o'`), and a dot-dash pattern (`'-.'`). Look through the matplotlib
-gallery for more examples.
+By default, matplotlib creates a figure in a separate window. When using
```
@wrightaprilm, I think this note about `%matplotlib inline` should be a callout box and introduced after the first plot is created.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-A single figure can include multiple lines, and they can be plotted using the
-same `plt.plot()` command by adding more pairs of x values and y values (and
-optionally line styles):
+Matplotlib integrates nicely with the `numpy` package and can use numpy arrays as input of the available plot functions. Consider the following example data, created with `numpy`:
```
> Matplotlib works seamlessly

maxim-belkin replied:
Note how we write Matplotlib and numpy. We need agree on how we write all the library names: `library` or Library.

I think it is fine either way, it just has to be consistent. Perhaps, Numpy and Matplotlib ?

wrightaprilm replied:
I addressed pretty much all the comments except this one. I could have sworn matplotlib was never capitalized, but it looks in their documentation like it is, now. It looks like NumPy has also transitioned to the "Py-emphasizing" case. pandas is still lowercase. So I say we go with how the developers do it - Matplotlib, NumPy, pandas.

Thoughts?

maxim-belkin replied:
I agree

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
-    plt.plot(t, t**3, 'g^:', label='cubic')
-
-    plt.legend(loc='upper left', shadow=True, fontsize='x-large')
+or create a matplotlib `figure` and `axis` object first and add the plot to the created objects:
```
> or create matplotlib `figure` and `axis` objects first and add the plot to these objects later on:

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
+![line plot](../fig/08_line_plot.png)
 
- The functions `xlabel`, `ylabel`, `title`, `legend`, and many others create text labels. It is good to know that, in addition to the plain text, you may use mathematical notation using a subset of LaTeX language. See [this link](https://matplotlib.org/users/mathtext.html) for more information.
+Although the latter requires a little bit more code to create the same plot, the advantage is that we now have **full control** of where the plot axis are placed, and we can easily add new items or, for example, more than one axis to the figure and adapt the labels:
```
Although the latter approach

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
+ax1.set_ylabel('y')
+ax1.set_xlabel('x')
 
+# add an additional ax to the figure
```
> additional axis

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
+### Link matplotlib, Pandas and plotnine
 
-    plt.show()
+The **Pandas and plotnine packages create matplotlib objects** as well. Hence, using the object based approach provides a consistent workflow and interaction between these packages:
```
I suggest replacing the two sentences with a text based on the following paragraph:

> When we tell Pandas or plotnine to create a plot, both libraries, in fact, turn to matplotlib to do the task and both libraries create matplotlib plots, not their own. Internally, Python stores these plots as matplotlib "objects" with a  bunch of predefined operations that we can apply to those. For example, set a label for an axis or add a title, or anything else that has to do with the plot.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
+surveys.plot("hindfoot_length", "weight", kind="scatter", ax=ax1)
 
-    plt.show()
+# Provide further adaptations with matplotlib:
```
This is not a good example for a comment, IMHO. Comments should be specific to the code, so here it should be something like `Set a label for the x axis`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,226 @@ also change the name of the columns in the DataFrame like this:
 
-    plt.subplot(2,2,3)  # Two row, two columns, position 3
-    plt.plot(t, t**3, 'g^:', label='cubic')
+Once satisfied about the resulting plot, you can save the plot with the `.savefig(*args)` command from matplotlib:
```
Should it be `Once satisfied with ....`?

stijnvanhoey commented:
@maxim-belkin and @wrightaprilm, are there any remaining elements that need to be adressed?

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,234 @@ also change the name of the columns in the DataFrame like this:
+> when working at the command line, with a script, or at the iPython interpreter. In the 
+> previous example, adding  `plt.show()` after the creation of the plot will enable your 
+> colleagues who aren't using a Jupyter notebook to reproduce your work on their platform.
+{: .callout}
```
How about moving this callout box to right after when `plt.plot` command is used and before the `Line plot of x vs y`

wrightaprilm replied:
I tried that and it looks a bit busy. How about the way I have it now?

maxim-belkin replied:
looks good!

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
 keypoints:
-    - "FIXME"
+    - "Matplotlib is the engine behind plotnine and Pandas plots."
+    - "The object-oriented framework of matplotlib makes detailed customization possible."
```
I'm not sure how I feel about this keypoint. What would attendees get out of it? Keypoints should be "to the point", that is, related to actions attendees learn during the lesson.

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,235 @@ also change the name of the columns in the DataFrame like this:
-Matplotlib can make many other types of plots in much the same way that it makes
-2 dimensional line plots. Look through the examples in
-http://matplotlib.org/users/screenshots.html and try a few of them (click on the
+Matplotlib can make many other types of plots in much the same way that it makes 2 dimensional line plots. Look through the examples in
```
two-dimensional

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,235 @@ also change the name of the columns in the DataFrame like this:
-> using the streamgage file, you could make a histogram of the number of days with
-> a given mean discharge, use bar plots to display daily discharge statistics, or
-> explore the different ways matplotlib can handle dates and times for figures.
+> using the streamgage file `bouldercreek_09_2013.txt`, you could make a 
```
perhaps, add a link to the file `[bouldercreek_09_2013.txt]({{ page.root }}/data/bouldercreek_09_2013.txt)`

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,235 @@ also change the name of the columns in the DataFrame like this:
 
-    plt.subplot(2,2,3)  # Two row, two columns, position 3
-    plt.plot(t, t**3, 'g^:', label='cubic')
+Once satisfied with the resulting plot, you can save the plot with the `.savefig(*args)` command from matplotlib:
```
Regarding command vs. method: do we want to be accurate here or do we want to sacrifice correctness for clarity?

stijnvanhoey commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,234 @@ also change the name of the columns in the DataFrame like this:
-figure number inside the parentheses:
+When we create a plot using pandas or plotnine, both libraries use matplotlib 
+to create those plots. The plots created in pandas or plotnine are matplotlib 
+objects, which enables us to use some of the advanced plotting options available 
```
I would keep them as they are as it explains the connection a bit more than just stating *the objects are matplotlib*. Maybe the second sentence can be removed, if this would improve the flow of the text. As you prefer, feel free to remove

stijnvanhoey commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -86,209 +89,235 @@ also change the name of the columns in the DataFrame like this:
+to create those plots. The plots created in pandas or plotnine are matplotlib 
+objects, which enables us to use some of the advanced plotting options available 
+in the matplotlib library. Because the objects output by pandas and plotnine
+can be read by matplotlib, we have many more options than any one library can 
```
See the previous comment, feel free to remove if required.

stijnvanhoey commented:
I tried to cover the remaining comments as far as I could.

maxim-belkin commented:
Thanks, @stijnvanhoey. I made minor changes.
The last thing I suggest we do in this PR before merging it is update time estimates for the episode

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -63,232 +67,264 @@ measurements. Convert any measurements in imperial units into SI units. You can
-point (`'o'`), and a dot-dash pattern (`'-.'`). Look through the matplotlib
-gallery for more examples.
+
+### `plt` pyplot versus object-based matplotlib
```
I'm not sure I understand the "comparison" (implied by "versus") here...

maxim-belkin replied:
we import `pyplot` module of `matplotlib` package as `plt` (`plt === matplotlib.pyplot`).
@stijnvanhoey, what were you trying to convey with this title?

...`layered plots`... have not been defined yet :)

stijnvanhoey replied:
I actually got *versus* from the matplotlib documentation itself: https://matplotlib.org/tutorials/introductory/lifecycle.html#a-note-on-the-object-oriented-api-vs-pyplot 

Using *`plt` pyplot* to refer to what learners will typically recognise already or encounter online (e.g. stackoverflow) instead of adding more technical terms such as *state-based interface*...

maxim-belkin replied:
Oh, that! That's pretty "deep"!  I'd suggest moving this discussion (about stateful `plt` vs OO matplotlib) to a callout box or something (or drop entirely). Or add to instructors notes so that at least instructors are aware of such a thing.

stijnvanhoey replied:
Maybe we should create a separate issue on how to tackle this with respect to squashing/merging?

maxim-belkin commented on _episodes/08-putting-it-all-together.md:
```diff
@@ -7,8 +7,11 @@ questions:
 keypoints:
-    - "FIXME"
+    - "Matplotlib is the engine behind plotnine and Pandas plots."
+    - "Object-based nature of matplotlib plots enables their detailed customization after they have been created."
```
`Modular` and `Object-oriented` are different beasts. Modular means one can bring together different pieces of something and they will work together. Object-oriented means that after we strip out all the complexities, we're dealing with (some) objects that have certain attributes and  methods associated with them. I think here we want to talk about Object-oriented side of things

stijnvanhoey replied:
I do agree. This is about matplotlib objects (with attributes and methods) and not about modularity.

wrightaprilm commented:
I'm going to run this tomorrow in my computational biology class, and will have a time estimate then.

maxim-belkin commented:
Will you be using a locally-rendered version of the website?

wrightaprilm commented:
I'll be converting it to a jupyter notebook and serving it on my jupyter hub. So I'll Stijn's fork and build from there.

wrightaprilm commented:
Adds about 1:45. I'm counting from this episode functionally not existing right now - not sure how long what is there right now takes. I'm in favor of merge; I'll make an updated schedule once that's done.

maxim-belkin commented:
Sounds good. Let's squash and merge this PR and update timings in a separate PR

wrightaprilm commented:
Squashed and merged. Thanks so much, both of you, this is a really nice addition. I might make some edits in the coming weeks, according to sticky points I saw in the run-through, but I'll tag you both in when I get it done.

Very exciting stuff!

maxim-belkin commented:
Excellent work, @stijnvanhoey!

stijnvanhoey commented:
Thanks for the in-depth reviewing @maxim-belkin, certainly improved the material.
